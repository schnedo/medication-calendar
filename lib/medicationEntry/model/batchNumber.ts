export default interface BatchNumber {
  readonly number: number;
}

const prefix = "Ch.-B.";
export function format(batchNumber: BatchNumber): string {
  return `${prefix} ${batchNumber.number}`;
}
