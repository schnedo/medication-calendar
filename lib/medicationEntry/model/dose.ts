export default interface Dose {
  readonly amount: number;
}

const unit = "ml";

export function format(dose: Dose): string {
  return `${dose.amount} ${unit}`;
}
