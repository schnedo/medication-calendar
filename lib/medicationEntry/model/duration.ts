export default interface Duration {
  minutes: number;
  hours: number;
}

function formatPart(number: number | undefined) {
  if (number === undefined) {
    return "00";
  }
  const raw = number.toString();
  return raw.length === 1 ? `0${raw}` : raw;
}

export function format(duration: Duration): string {
  return `${formatPart(duration.hours)}:${formatPart(duration.minutes)}`;
}
