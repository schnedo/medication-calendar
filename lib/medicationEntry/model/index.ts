export { default as BatchNumber } from "./batchNumber";
export { default as Dose } from "./dose";
export { default as Duration } from "./Duration";
export { default as Medicament } from "./medicament";
export { default as Medication } from "./medication";
export { default as MedicationEntry } from "./medicationEntry";

export * as Repository from "./repository";
