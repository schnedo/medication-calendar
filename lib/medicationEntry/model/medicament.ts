import BatchNumber from "./batchNumber";

export default interface Medicament {
  readonly name: string;
  readonly batchNumber: BatchNumber;
}
