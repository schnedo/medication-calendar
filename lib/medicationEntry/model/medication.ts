import Dose from "./dose";
import Medicament from "./medicament";
import { v4 } from "uuid";

export default class Medication {
  readonly id: string = v4();

  constructor(readonly dose: Dose, readonly medicament: Medicament) {}
}
