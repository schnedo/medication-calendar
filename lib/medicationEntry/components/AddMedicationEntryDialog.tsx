import { MouseEventHandler, ReactElement, useState } from "react";
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  makeStyles,
  TextField,
} from "@material-ui/core";
import { DatePicker } from "@material-ui/pickers";
import { ModalProps } from "@material-ui/core/Modal";
import { Duration, Medication, MedicationEntry } from "../model";
import DurationInput from "./DurationInput";
import MedicationsInput from "./MedicationsInput";

const useStyles = makeStyles((theme) => ({
  dialogContent: {
    display: "flex",
    flexDirection: "column",
    "&>*": {
      marginBottom: theme.spacing(2),
      "&:lastChild": {
        marginBottom: theme.spacing(0),
      },
    },
  },
}));

export interface AddMedicationEntryDialogProps {
  open: boolean;
  onClose?: ModalProps["onClose"];
  onSubmit?: (entry: MedicationEntry) => Promise<void>;
  onAbort?: () => void;
}

export default function AddMedicationEntryDialog({
  onClose,
  onSubmit,
  onAbort,
  open,
}: AddMedicationEntryDialogProps): ReactElement {
  const { dialogContent } = useStyles();

  const [date, setDate] = useState(new Date());
  const [duration, setDuration] = useState({
    minutes: 0,
    hours: 0,
  } as Duration);
  const [bodyMass, setBodyMass] = useState("60");
  const [comments, setComments] = useState("");
  const [medications, setMedications] = useState<Medication[]>([]);

  const handleSubmit: MouseEventHandler | undefined = onSubmit
    ? async () => {
        await onSubmit({
          id: "",
          date,
          duration,
          bodyMass: {
            amount: parseInt(bodyMass),
          },
          comments,
          medications,
        });
      }
    : undefined;

  return (
    <Dialog open={open} onClose={onClose}>
      <DialogTitle>Neuer Eintrag</DialogTitle>
      <DialogContent className={dialogContent}>
        <TextField
          id={"bodyMassInput"}
          label={"KÃ¶rpergewicht"}
          type={"number"}
          value={bodyMass}
          onChange={(event) => setBodyMass(event.target.value)}
        />
        <DatePicker
          id={"dateInput"}
          label={"Datum"}
          value={date}
          onChange={(value) => setDate(value ?? date)}
        />
        <DurationInput value={duration} onChange={setDuration} />
        <TextField
          id={"commentsInput"}
          label={"Kommentare"}
          multiline
          value={comments}
          onChange={(event) => setComments(event.target.value)}
        />
        <MedicationsInput value={medications} onChange={setMedications} />
      </DialogContent>
      <DialogActions>
        <Button color="primary" onClick={onAbort}>
          Abbrechen
        </Button>
        <Button color="primary" onClick={handleSubmit}>
          Speichern
        </Button>
      </DialogActions>
    </Dialog>
  );
}
