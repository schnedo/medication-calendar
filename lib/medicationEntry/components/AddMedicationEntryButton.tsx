import { ReactElement, useState } from "react";
import { Fab, makeStyles } from "@material-ui/core";
import { Add } from "@material-ui/icons";
import AddMedicationEntryDialog from "./AddMedicationEntryDialog";
import { useMedicationEntries } from "./MedicationEntriesProvider";
import { MedicationEntry } from "../model";

const useStyles = makeStyles((theme) => ({
  actionButton: {
    position: "fixed",
    bottom: theme.spacing(2),
    right: theme.spacing(2),
  },
}));

export default function AddMedicationEntryButton(): ReactElement {
  const { actionButton } = useStyles();

  const [open, setOpen] = useState(false);
  const showDialog = () => setOpen(true);
  const closeDialog = () => setOpen(false);
  const { saveEntry } = useMedicationEntries();
  const onSubmit = async (entry: MedicationEntry) => {
    await saveEntry(entry);
    closeDialog();
  };

  return (
    <>
      <Fab
        color="primary"
        aria-label="add"
        onClick={showDialog}
        className={actionButton}
      >
        <Add />
      </Fab>
      <AddMedicationEntryDialog
        open={open}
        onSubmit={onSubmit}
        onAbort={closeDialog}
        onClose={closeDialog}
      />
    </>
  );
}
