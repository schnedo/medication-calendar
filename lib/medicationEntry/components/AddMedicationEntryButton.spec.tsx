import { mock } from "@userlike/joke";
import { render, waitFor } from "@testing-library/react";
import AddMedicationEntryButton from "./AddMedicationEntryButton";
import userEvent from "@testing-library/user-event";
import { MedicationEntry } from "../model";

const { default: AddMedicationEntryDialog } = mock(
  import("./AddMedicationEntryDialog"),
);
const { useMedicationEntries } = mock(import("./MedicationEntriesProvider"));

describe("AddMedicationEntryButton", () => {
  beforeEach(() => {
    AddMedicationEntryDialog.mockReturnValue(
      <div>mocked AddMedicationEntryDialog</div>,
    );
    useMedicationEntries.mockReturnValue({
      entries: [],
      saveEntry: jest.fn().mockResolvedValue(undefined),
      deleteEntry: jest.fn().mockResolvedValue(undefined),
    });
  });

  it("should render correctly", async () => {
    expect.hasAssertions();

    const { container } = render(<AddMedicationEntryButton />);

    expect(container).toMatchSnapshot();
  });

  it("should open AddMedicationEntryDialog on press", async () => {
    expect.hasAssertions();

    const { getByLabelText } = render(<AddMedicationEntryButton />);

    expect(AddMedicationEntryDialog).toHaveBeenCalledTimes(1);
    expect(AddMedicationEntryDialog).toHaveBeenLastCalledWith(
      expect.objectContaining({ open: false }),
      {},
    );
    userEvent.click(getByLabelText("add"));
    await waitFor(() => {
      expect(AddMedicationEntryDialog).toHaveBeenCalledTimes(2);
    });
    expect(AddMedicationEntryDialog).toHaveBeenLastCalledWith(
      expect.objectContaining({ open: true }),
      {},
    );
  });

  it("should save entry returned by dialog submit and close dialog", async () => {
    expect.hasAssertions();

    const entry: MedicationEntry = {
      bodyMass: {
        amount: 60,
      },
      comments: "comments",
      date: new Date(2021, 1, 10),
      duration: { minutes: 15, hours: 2 },
      id: "",
      medications: [],
    };
    AddMedicationEntryDialog.mockImplementation(({ onSubmit }) => (
      <div data-testid={"dialog"} onClick={() => onSubmit && onSubmit(entry)} />
    ));
    const saveEntry = jest.fn();
    useMedicationEntries.mockReturnValue({
      entries: [],
      saveEntry,
      deleteEntry: jest.fn(),
    });

    const { getByLabelText, getByTestId } = render(
      <AddMedicationEntryButton />,
    );

    userEvent.click(getByLabelText("add"));
    await waitFor(() => {
      expect(AddMedicationEntryDialog).toHaveBeenCalledTimes(2);
    });
    userEvent.click(getByTestId("dialog"));
    await waitFor(() => {
      expect(AddMedicationEntryDialog).toHaveBeenCalledTimes(3);
    });
    expect(saveEntry).toHaveBeenCalledTimes(1);
    expect(saveEntry).toHaveBeenLastCalledWith(entry);
    expect(AddMedicationEntryDialog).toHaveBeenLastCalledWith(
      expect.objectContaining({ open: false }),
      {},
    );
  });

  it("should close Dialog when AddMedicationEntryDialog calls onClose", async () => {
    expect.hasAssertions();

    AddMedicationEntryDialog.mockImplementation(({ onClose }) => (
      <div
        data-testid={"dialog"}
        onClick={() => onClose && onClose({}, "backdropClick")}
      />
    ));

    const { getByLabelText, getByTestId } = render(
      <AddMedicationEntryButton />,
    );

    userEvent.click(getByLabelText("add"));
    userEvent.click(getByTestId("dialog"));
    expect(AddMedicationEntryDialog).toHaveBeenLastCalledWith(
      expect.objectContaining({ open: false }),
      {},
    );
  });

  it("should close dialog when AddMedicationEntryDialog calls onAbort", async () => {
    expect.hasAssertions();

    AddMedicationEntryDialog.mockImplementation(({ onAbort }) => (
      <div data-testid={"dialog"} onClick={() => onAbort && onAbort()} />
    ));

    const { getByLabelText, getByTestId } = render(
      <AddMedicationEntryButton />,
    );

    userEvent.click(getByLabelText("add"));
    userEvent.click(getByTestId("dialog"));
    expect(AddMedicationEntryDialog).toHaveBeenLastCalledWith(
      expect.objectContaining({ open: false }),
      {},
    );
  });
});
