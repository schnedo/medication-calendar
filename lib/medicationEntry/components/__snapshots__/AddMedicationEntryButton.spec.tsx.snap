// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMedicationEntryButton should render correctly 1`] = `
<div>
  <button
    aria-label="add"
    class="MuiButtonBase-root MuiFab-root makeStyles-actionButton-1 MuiFab-primary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <div>
    mocked AddMedicationEntryDialog
  </div>
</div>
`;
